test/modules/TwinTester/DesiredPropertiesValidator.cs

```diff
@@ -56,20 +62,20 @@ public async Task ValidateAsync()
             foreach (KeyValuePair<string, DateTime> desiredPropertyUpdate in desiredPropertiesUpdated)
             {
                 bool hasTwinUpdate = propertyUpdatesFromTwin.Contains(desiredPropertyUpdate.Key);
-                bool hasModuleReceivedCallback = desiredPropertiesReceived.ContainsKey(desiredPropertyUpdate.Key);
+                bool hasReceivedPotentialCallback = desiredPropertiesReceived.ContainsKey(desiredPropertyUpdate.Key) || !this.ShouldExpectDesiredPropertyCallback(this.twinState, desiredPropertyUpdate.Value);
                 string status;
-                if (hasTwinUpdate && hasModuleReceivedCallback)
+                if (hasTwinUpdate && hasReceivedPotentialCallback)
```

Comment:
1. desiredPropertyUpdate.key will miss some keys due to edgeHub restart? Why do we  use it? will we miss some key to validate?
2. what if twin update and callback both return true here, but it takes longer than failure threshold?  should this result be treated as fail?

Reply:
1 - Do you mean the desiredPropertyUpdate.key will increment by more than one? We update the twin with a specific key so we can match it up when validating.
2 - The logic is more forgiving and tries to validate potentially past the failure threshold. This is acceptable because the logic runs on a timer (PeriodicTask), therefore this update could have potentially come within the failure threshold but is simply being validated at a later time. If the logic wasn't forgiving then it would fail this valid case.

Reply:
1. you are right.
2. If both send and receive will stamp a timestamp, then it should be good enough for comparison and achieve your goal to make sure "all twin should be received within failure threshold".  I am open to this, your call.

Reply:
2 - I think your idea is slightly better, but don't think we should prioritize it now.